pkg/bundle/builder.go

```diff
@@ -0,0 +1,315 @@
+	glog.Infof("building bundle for %q", ig.Name)
+	keyStore, err := b.Clientset.KeyStore(cluster)
+	if err != nil {
+		return nil, err
+	}
+
+	fullCluster := &kops.Cluster{}
+	{
+		configBase, err := b.Clientset.ConfigBaseFor(cluster)
+		if err != nil {
+			return nil, fmt.Errorf("error building ConfigBase for cluster: %v", err)
+		}
+
+		p := configBase.Join(registry.PathClusterCompleted)
+
+		b, err := p.ReadFile()
+		if err != nil {
+			return nil, fmt.Errorf("error loading Cluster %q: %v", p, err)
+		}
+
+		err = utils.YamlUnmarshal(b, fullCluster)
+		if err != nil {
+			return nil, fmt.Errorf("error parsing Cluster %q: %v", p, err)
+		}
+	}
+
+	glog.Infof("fullCluster %v", fullCluster)
+
+	fullCluster.Spec.ConfigBase = "/etc/kubernetes/bootstrap"
+	fullCluster.Spec.ConfigStore = "/etc/kubernetes/bootstrap"
+	fullCluster.Spec.KeyStore = "/etc/kubernetes/bootstrap/pki"
+	fullCluster.Spec.SecretStore = "/etc/kubernetes/bootstrap/secrets"
+
+	var files []*DataFile
+
+	{
+		data, err := utils.YamlMarshal(fullCluster)
+		if err != nil {
+			return nil, fmt.Errorf("error marshalling configuration: %v", err)
+		}
+
+		file := &DataFile{}
+		file.Header.Name = "cluster.spec"
+		file.Header.Size = int64(len(data))
+		file.Header.Mode = 0644
+		file.Data = data
+		files = append(files, file)
+	}
+
+	{
+		data, err := kopscodecs.ToVersionedYaml(ig)
+		if err != nil {
+			return nil, fmt.Errorf("error encoding instancegroup: %v", err)
+		}
+
+		file := &DataFile{}
+		file.Header.Name = "instancegroup/" + ig.Name
+		file.Header.Size = int64(len(data))
+		file.Header.Mode = 0644
+		file.Data = data
+		files = append(files, file)
+	}
+
+	if pkiFiles, err := b.buildPKIFiles(cluster, ig, keyStore); err != nil {
+		return nil, err
+	} else {
+		glog.Infof("pki files %v", pkiFiles)
+		files = append(files, pkiFiles...)
+	}
+
+	copyManifest := make(map[string]string)
+
+	{
+		phase := cloudup.PhaseCluster
+		assetBuilder := assets.NewAssetBuilder(cluster.Spec.Assets, string(phase))
+
+		applyCmd := &cloudup.ApplyClusterCmd{
+			Cluster:        cluster,
+			Clientset:      b.Clientset,
+			InstanceGroups: []*kops.InstanceGroup{ig},
+			Phase:          phase,
+		}
+
+		if err := applyCmd.AddFileAssets(assetBuilder); err != nil {
+			return nil, fmt.Errorf("error adding assets: %v", err)
+		}
+
+		nodeupConfig, err := applyCmd.BuildNodeUpConfig(assetBuilder, ig)
+		if err != nil {
+			return nil, fmt.Errorf("error building nodeup config: %v", err)
+		}
+
+		if !ig.IsMaster() {
+			nodeupConfig.ProtokubeImage = nil
+			nodeupConfig.Channels = nil
+		}
+
+		nodeupConfig.ConfigBase = fi.String("/etc/kubernetes/bootstrap")
+
+		{
+			var localChannels []string
+			for _, channel := range nodeupConfig.Channels {
+				base := path.Base(channel)
+				localChannel := "file://" + path.Join("/rootfs/etc/kubernetes/bootstrap/addons/", base)
+				localChannels = append(localChannels, localChannel)
```

Comment:
Should we make a todo that channels needs to be externalized? Having channels pulled down makes more sense to me than bundling it.

Reply:
Not sure what you mean exactly, but yes, my plan is to split protokube into:

* kube-discovery
* etcd-manager
* channels (addon-manager)

We really only need channels on the master, and the bundle method is really just for nodes right now.  It would be great to download addons locally, but we still have to get them from somewhere :-)